runner/stats.go

```diff
@@ -20,64 +21,91 @@ const (
-type TestStats struct {
-	Run        int
-	Failed     []string
-	Skipped    []string
-	Ignored    []string
-	ForcedPass []string
-	ForcedFail []string
-	Success    int
-	RunTime    time.Duration
+// RunStats accumulates test statistics.
+type RunStats struct {
+	// Run is the amount of tests executed in this run.
+	Run int `json:"run"`
+	// Success is a list containing the tests that were successful.
+	Success []string `json:"success"`
+	// Failed is a list containing the failed tests.
+	Failed []string `json:"failed"`
+	// Skipped is a list containing the tests that were skipped.
+	Skipped    []string `json:"skipped"`
+	Ignored    []string `json:"ignored"`
+	ForcedPass []string `json:"forced-pass"`
+	ForcedFail []string `json:"forced-fail"`
+	// RunTime maps the time taken to run each test.
+	RunTime map[string]time.Duration `json:"runtime"`
+	// TotalTime is the total duration for the whole run.
+	TotalTime time.Duration
 }
 
-func (t *TestStats) TotalFailed() int {
-	return len(t.Failed) + len(t.ForcedFail)
+// NewRunStats creates and initializes a new Stats struct.
+func NewRunStats() *RunStats {
+	return &RunStats{
+		Run:        0,
+		Success:    nil,
+		Failed:     nil,
+		Skipped:    nil,
+		Ignored:    nil,
+		ForcedPass: nil,
+		ForcedFail: nil,
+		RunTime:    make(map[string]time.Duration),
+		TotalTime:  0,
+	}
+}
+
+func (stats *RunStats) TotalFailed() int {
+	return len(stats.Failed) + len(stats.ForcedFail)
 }
 
-func addResultToStats(result TestResult, title string, stats *TestStats) {
+func (stats *RunStats) addResultToStats(result TestResult, title string, testTime time.Duration) {
 	switch result {
 	case Success:
-		stats.Success++
+		stats.Success = append(stats.Success, title)
+		stats.RunTime[title] = testTime
 	case Failed:
 		stats.Failed = append(stats.Failed, title)
+		stats.RunTime[title] = testTime
 	case Skipped:
 		stats.Skipped = append(stats.Skipped, title)
 	case Ignored:
 		stats.Ignored = append(stats.Ignored, title)
 	case ForceFail:
 		stats.ForcedFail = append(stats.ForcedFail, title)
+		stats.RunTime[title] = testTime
 	case ForcePass:
 		stats.ForcedPass = append(stats.ForcedPass, title)
+		stats.RunTime[title] = testTime
 	default:
 		log.Info().Msgf("runner/stats: don't know how to handle TestResult %d", result)
 	}
 }
 
-func printSummary(quiet bool, stats TestStats) {
-	if quiet {
-		return
-	}
-
+func (stats *RunStats) printSummary(out *output.Output) {
 	if stats.Run > 0 {
-		emoji.Printf(":plus:run %d total tests in %s\n", stats.Run, stats.RunTime)
-		emoji.Printf(":next_track_button: skipped %d tests\n", len(stats.Skipped))
-		if len(stats.Ignored) > 0 {
-			emoji.Printf(":index_pointing_up: ignored %d tests\n", len(stats.Ignored))
-		}
-		if len(stats.ForcedPass) > 0 {
-			emoji.Printf(":index_pointing_up: forced to pass %d tests\n", len(stats.ForcedPass))
-		}
-		if stats.TotalFailed() == 0 {
-			emoji.Println(":tada:All tests successful!")
+		if out.Type() == "json" {
```

Comment:
This should be encapsulated in the output type somehow

Reply:
Agreed.

Reply:
Just created a new `IsJson` method 🤷